autoPyTorch/pipeline/components/setup/network/base_network.py

```diff
@@ -106,24 +111,35 @@ def predict(self, loader: torch.utils.data.DataLoader) -> torch.Tensor:
         """
         Performs batched prediction given a loader object
         """
-        assert self.network is not None
-        self.network.eval()
-
+        if len(self.network_snapshots) == 0:
+            assert self.network is not None
+            return self._predict(network=self.network, loader=loader).numpy()
+        else:
+            # if there are network snapshots,
+            # take average of predictions of all snapshots
+            Y_snapshot_preds: List[torch.Tensor] = list()
+
+            for network in self.network_snapshots:
+                Y_snapshot_preds.append(self._predict(network, loader))
+            Y_snapshot_preds_tensor = torch.stack(Y_snapshot_preds)
+            return Y_snapshot_preds_tensor.mean(dim=0).numpy()
+
+    def _predict(self, network: torch.nn.Module, loader: torch.utils.data.DataLoader) -> torch.Tensor:
+        network.to(self.device)
+        network.float()
+        network.eval()
         # Batch prediction
         Y_batch_preds = list()
 
         for i, (X_batch, Y_batch) in enumerate(loader):
             # Predict on batch
             X_batch = X_batch.float().to(self.device)
+            Y_batch_pred = network(X_batch)
+            if self.final_activation is not None:
+                Y_batch_pred = self.final_activation(Y_batch_pred)
+            Y_batch_preds.append(Y_batch_pred.detach().cpu())
 
-            with torch.no_grad():
-                Y_batch_pred = self.network(X_batch)
-                if self.final_activation is not None:
-                    Y_batch_pred = self.final_activation(Y_batch_pred)
-
-            Y_batch_preds.append(Y_batch_pred.cpu())
-
-        return torch.cat(Y_batch_preds, 0).cpu().numpy()
+        return torch.cat(Y_batch_preds, 0)
```

Comment:
Because now the predictions are already on CPU from line:
https://github.com/automl/Auto-PyTorch/blob/abeeaa4e8560b9736dc674d1e62ea98a1d094bdd/autoPyTorch/pipeline/components/setup/network/base_network.py#L140

Also additionally, they are being put to numpy in lines:

https://github.com/automl/Auto-PyTorch/blob/abeeaa4e8560b9736dc674d1e62ea98a1d094bdd/autoPyTorch/pipeline/components/setup/network/base_network.py#L116

https://github.com/automl/Auto-PyTorch/blob/abeeaa4e8560b9736dc674d1e62ea98a1d094bdd/autoPyTorch/pipeline/components/setup/network/base_network.py#L125